pilot/pkg/config/clusterregistry/conversion.go

```diff
@@ -43,14 +43,10 @@ const (
 	ClusterAccessConfigSecretNamespace = "config.istio.io/accessConfigSecretNamespace"
 )
 
-// Metadata defines a struct used as a key
-type Metadata struct {
```

Comment:
The reason for that is that the cluster name has nothing to do with namespaces. In the case of multiple secrets that contains cluster information, the keys inside the secrets should not overlap. Otherwise, they will be considered as referencing to the same clusters.

if I understand multiple istio in one cluster correctly, then it'll have multiple instances of pilot running. I don't see problem over there. I guess that in that case, an instance of pilot would only process secrets belonging to it. In other words, the secrets themselves need to be scoped properly.

Reply:
It's kind of related now that the data keys in the secrets are considered the cluster names and thus the keys in this map

Reply:
@baodongli What is the issue if keep using `[clustername, namespace]` as the key?

Reply:
@baodongli any comments for this question? The overall PR looks good to me except this question.

BTW: We are not able to do the e2e test because of issue https://github.com/istio/istio/issues/6418

Reply:
@gyliu513 to answer your question directly, I don't think there is an issue using [clustername, namespace] as a KEY. The question is whether or not it still makes sense to do that. Without this PR, it imposes the unnecessary restriction that the secret name has to be the same as the kubeconfig file name. So it probably makes sense to do that since the secret name is namespaced. With that restriction removed, it doesn't make sense to do that any more.

Reply:
Thanks @baodongli , that's also a question from me as well: still did not clear why pilot crash if I define a secret with a name different from kubeconfig name, can you please show more detail here?

I have a kubeconfig named as `cluster88` and created a secret named as `abc`, then pilot crashed.

In its log, I have the following:
```
2018-06-20T22:23:26.893113Z	info	Processing add: istio-system/abc
2018-06-20T22:23:26.893158Z	info	Adding new cluster member: abc
2018-06-20T22:23:26.893301Z	info	Service controller watching namespace "" for service, endpoint, nodes and pods, refresh 60000000000
2018-06-20T22:23:26.893942Z	info	Number of clusters in the cluster store: 1
```

The above means that the cluster has been added successfully, then pilot got crashed, the key now is `[abc, istio-system]`, why crashed?

Reply:
Ah, got it, caused by here https://github.com/istio/istio/blob/master/pilot/pkg/config/clusterregistry/secretcontroller.go#L215

Reply:
@baodongli I think even if I create a secret with a different name then the config file, the kubeconfig should still be loaded successfully?

/cc @sdake @sbezverk @tiswanso

Reply:
I think removing this does not make any sense,  you do not buy anything by removing it, but when istio will need to a capability of being installed into a different namespace, removal will cost you another refactor.

Reply:
@sbezverk I still not clear for the root cause of why pilot got crashed for this case, as I can see the pilot log does reporting that the remote cluster was added successfully before crash.
```
2018-06-20T22:23:26.893113Z	info	Processing add: istio-system/abc
2018-06-20T22:23:26.893158Z	info	Adding new cluster member: abc
2018-06-20T22:23:26.893301Z	info	Service controller watching namespace "" for service, endpoint, nodes and pods, refresh 60000000000
2018-06-20T22:23:26.893942Z	info	Number of clusters in the cluster store: 1
```

Reply:
@gyliu513 If you could get me the secret which causes the crash, I will get RCA, it seems @baodongli is already working on the fix, which more looks like a refactoring ;).

Reply:
@sbezverk , it is pretty simple to reproduce: suppose you have a kubeconfig named as `a`, then you create a secret name as `b` and label it with multicluster flag, pilot will got crashed.

Reply:
@gyliu513 Sorry that I thought that it was clear why it crashed. But this is why: if the secret name is not the same as the filenames that are used to create the secret, then the map lookup will return a "" string, in which case, clientcmd.load() still succeeds. but the client config returned can't be used as it's later in the code (in secret controller.go). The same happens when the secret contain an empty kubeconfig file.

Reply:
I see

```
root@gyliu-c21:~/multicluster# kubectl get secrets -n istio-system abc -oyaml
apiVersion: v1
data:
  cluster88: xxxx <<<<
kind: Secret
metadata:
  creationTimestamp: 2018-06-21T22:16:55Z
  labels:
    istio/multiCluster: "true"
  name: abc <<<< 
  namespace: istio-system
  resourceVersion: "164498"
  selfLink: /api/v1/namespaces/istio-system/secrets/abc
  uid: ce61cb3b-75a0-11e8-b0d7-00163e01fa8d
type: Opaque
```